sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/InternalColumnarRddConverter.scala

```diff
@@ -280,90 +437,133 @@ private object GpuExternalRowToColumnConverter {
+    extends TypeConverter {
+    override def append(row: Row,
+      column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      if (row.isNullAt(column)) {
+        builder.appendNull()
+        VALIDITY_N_OFFSET
+      } else {
+        arrayConvert(childConverter, row, column, builder) + VALIDITY
+      }
+    }
+
+    override def getNullSize: Double = OFFSET + VALIDITY
+  }
+
+  private case class NotNullArrayConverter(childConverter: TypeConverter)
+    extends TypeConverter {
+    override def append(row: Row,
+      column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      arrayConvert(childConverter, row, column, builder)
+    }
+
+    override def getNullSize: Double = OFFSET + VALIDITY
+  }
+
+  private[this] def structConvert(
+    childConverters: Array[TypeConverter],
+    row: Row,
+    column: Int,
+    builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder) : Double = {
+    var ret = 0.0
+    val struct = row.getStruct(column)
+    for (i <- childConverters.indices) {
+      ret += childConverters(i).append(struct, i, builder.getChild(i))
+    }
+    builder.endStruct()
+    ret
+  }
+
+  private case class StructConverter(
+    childConverters: Array[TypeConverter]) extends TypeConverter {
+    override def append(row: Row,
+      column: Int,
+      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      if (row.isNullAt(column)) {
+        builder.appendNull()
+        childConverters.map(_.getNullSize).sum + VALIDITY
+        // each child has to insert a null too, which is dependent on the child
+      } else {
+        structConvert(childConverters, row, column, builder) + VALIDITY
+      }
+    }
+
+    override def getNullSize: Double = childConverters.map(_.getNullSize).sum + VALIDITY
+  }
+
+  private case class NotNullStructConverter(
+    childConverters: Array[TypeConverter]) extends TypeConverter {
+    override def append(row: Row,
+      column: Int,
+      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      structConvert(childConverters, row, column, builder)
+    }
+
+    override def getNullSize: Double = childConverters.map(_.getNullSize).sum + VALIDITY
+  }
+
+  private class DecimalConverter(
+    precision: Int, scale: Int) extends NotNullDecimalConverter(precision, scale) {
+    private val appendedSize = if (precision <= Decimal.MAX_LONG_DIGITS) {
+      8 + VALIDITY
+    } else {
+      16 + VALIDITY
+    }
+
+    override def append(
+      row: Row,
+      column: Int,
+      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      if (row.isNullAt(column)) {
+        builder.appendNull()
+      } else {
+        super.append(row, column, builder)
+      }
+      appendedSize
+    }
+  }
+
+  private class NotNullDecimalConverter(precision: Int, scale: Int) extends TypeConverter {
+    private val appendedSize = if (precision <= Decimal.MAX_INT_DIGITS) {
```

Comment:
See above comment, probably should leverage `DecimalUtil` rather than replicate logic here.